crates/redstone/src/helpers/trie.rs

```diff
@@ -0,0 +1,174 @@
+use crate::network::error::Error;
+use crate::FeedId;
+
+const CHAR_START: u8 = 48; // Start at ASCII b'0'
+const CHAR_END: u8 = 90; // End at ASCII b'Z'
+const CHARS_COUNT: usize = (CHAR_END - (CHAR_START - 1)) as usize;
+
+#[derive(Debug, Clone)]
+struct Node {
+    children: [usize; CHARS_COUNT],
+    end: bool,
+}
+
+impl Default for Node {
+    #[inline]
+    fn default() -> Self {
+        Self {
+            children: [0usize; CHARS_COUNT],
+            end: false,
+        }
+    }
+}
+
+/// The Trie data structure is a tree-like data structure used for storing a dynamic set of strings.
+/// It is commonly used for efficient retrieval and storage of keys in a large dataset.
+///
+/// Trie stores unique FeedIds to ensure no repetition of the same feed in given datastructure.
+#[derive(Debug, Clone)]
+pub struct Trie {
+    nodes: Vec<Node>,
+}
+
+impl Default for Trie {
+    fn default() -> Self {
+        let nodes = vec![Node::default()];
+        Self { nodes }
+    }
+}
+
+macro_rules! is_pattern {
+    ($obj:expr, $($matcher:pat),*) => {
+       match $obj {
+           $($matcher => true),*,
+            _ => false,
+       }
+   }
+}
+
+impl Trie {
+    /// Stores the feed_id in Trie and indicates is the feed_id a newly inserted value.
+    ///
+    /// The purpose of this function is to indicate if given feed_id already exist in the Trie datastructure,
+    /// which offers then to validate the uniqueenes of given feed_id.
+    ///
+    /// # Arguments
+    ///
+    /// * `feed_id` - A FeedId value.
+    /// It is acceptable for FeedId to contain ASSCI numbers and uppercase letters.
+    /// All characters from the list are allowed: [b'0', ..., b'9', b'A' ..., b'Z'].
+    ///
+    /// # Returns
+    ///
+    /// Returns a `Result<bool, Error>`, which indicates if given FeedId.
+    /// If The Trie contains FeedId already 'true' is returned, otherwise function returns 'false'.
+    pub fn store(&mut self, feed_id: &FeedId) -> Result<bool, Error> {
+        let mut cur_node = 0;
+
+        for c in feed_id.0.iter() {
+            if *c == b'\0' {
+                break;
+            }
+            if is_pattern!(*c, b':', b';', b'<', b'=', b'>', b'?', b'@')
```

Comment:
suggestion: 
Instead check if the c is in the range:
```rust
#[inline]
// check if the character is one of ':;<=>?@'
fn is_special_char(character: u8) -> bool {
    c >= 58 && c <= 64
}
```

Reply:
minor: extract the whole if condition to unction, for example named: `is_allowed_char`